plugins/module_utils/network/vyos/config/firewall_rules/firewall_rules.py

```diff
@@ -114,6 +114,17 @@ def execute_module(self):
             result["gathered"] = changed_firewall_rules_facts
 
         result["warnings"] = warnings
+
+        if self.state in self.ACTION_STATES and self._module._diff:
+            result["diff"] = {
+                "after": "\n".join(
+                    sorted(self.set_state(self._module.params["config"], {}))
+                ),
+                "before": "\n".join(
+                    sorted(self.set_state(existing_firewall_rules_facts, {}))
+                ),
+            }
```

Comment:
This is only the right thing for `overridden`, so this PR needs more work.  However, this got me thinking that there are two ways to accomplish this.

## Commands [to be] Run

You don't get to see the commands this module (returned via `results["commands"]`) unless you run your play with `-vvv` as far as I can tell.  That has *a lot* of other output, which makes it not useful in a lot of situations.  However, we could just include those as the "after" lines in a diff, and this change would then be quite simple.  It would amount to something like this:
```python
         if self.state in self.ACTION_STATES and self._module._diff:
            result["diff"] = {
                "after": "\n".join(result["commands"]),
                "before": "",
            }
```

## Resulting Config

This would basically generate what `show configuration commands` for the current state (easy) as well as the new state (hard).  This would require new helper methods for `deleted`, `merged`, and `replaced` states that will compute what the config object would look like once all the changes we want to apply would look like.  Ideally, we'd be able to share code between `_state_deleted`, `_state_merged`, and `_state_replaced`, but I don't actually see a reasonable way to do that.

## Comparison

| Option | Pros | Cons |
| --- | --- | --- |
| Commands Run | <ul><li>Easy to implement</li><li>Easy to test</li><li>Always accurate as to what was run</li></ul> | <ul><li>Not really what the diff command is meant to convey</li></ul> |
| Resulting Config | <ul><li>Accurate representation of changed sate on the device</li></ul> | <ul><li>Hard to implement</li><li>Hard to test</li><li>Risk of being inaccurate due to divergent code paths</li></ul> |

Despite the table outcome, I actually think the resulting config approach is the better one long-term.  However, I don't want to do a bunch of work to implement that if it's not wanted.  Therefore, I'm going to wait for some input from y'all before I push this further.

I'd also like to add some tests for this, but unfortunately `ansible-test` fails a bunch of tests for me, and there aren't really any docs that tell me how to run them locally for this module (at least that I've found).  Uploading and waiting for zuul to find out if it works or not isn't a way I want to develop this.

Reply:
Code re-use is actually possible here, so that's not a con of the resulting config approach after all.  If we go that route, I'll do a refractor in a seperate PR.

Reply:
@ashwini-mhatre, any thoughts here?  I don't want to do a bunch of work that the project doesn't ultimately want, but I think getting check mode added will be a valuable feature.

Reply:
@sdwilsh This seems to be a good feature in my mind as of now. Let me take 2 days to process this and discuss the same with the team and get back to you. Thank you.

Reply:
@GomathiselviS, any thoughts on direction to go here?

Reply:
@GomathiselviS, bump for any thoughts on direction to go here.  I'm still interested in solving this because I think it'll help track down an issue where I see my firewall rules always being listed as changed when I am pretty sure they are not.